plugins/reindex/src/main/java/org/elasticsearch/plugin/reindex/AbstractAsyncBulkByScrollAction.java

```diff
@@ -0,0 +1,352 @@
+                    }
+
+                    @Override
+                    public void onFailure(Throwable e) {
+                        finishHim(e);
+                    }
+                });
+            } catch (Throwable t) {
+                finishHim(t);
+            }
+        });
+    }
+
+    void onBulkResponse(BulkResponse response) {
+        try {
+            Set<String> destinationIndicesThisBatch = new HashSet<>();
+            for (BulkItemResponse item : response) {
+                if (item.isFailed()) {
+                    recordFailure(item.getFailure());
+                    continue;
+                }
+
+                switch (item.getOpType()) {
+                case "index":
+                case "create":
+                    IndexResponse ir = item.getResponse();
+                    if (ir.isCreated()) {
+                        created.incrementAndGet();
+                    } else {
+                        updated.incrementAndGet();
+                    }
+                    break;
+                case "delete":
+                    deleted.incrementAndGet();
+                    break;
+                default:
+                    throw new IllegalArgumentException("Unknown op type:  " + item.getOpType());
+                }
+                // Track the indexes we've seen so we can refresh them if requested
+                destinationIndicesThisBatch.add(item.getIndex());
+            }
+            if (destinationIndicesThisBatch.isEmpty() == false) {
+                destinationIndices.addAll(destinationIndicesThisBatch);
+            }
+
+            if (failures.isEmpty() == false) {
+                startNormalTermination();
+                return;
+            }
+
+            if (mainRequest.size() != -1 && successfullyProcessed() >= mainRequest.size()) {
+                // We've processed all the requested docs.
+                startNormalTermination();
+                return;
+            }
+            startNextScrollRequest();
+        } catch (Throwable t) {
+            finishHim(t);
+        }
+    }
+
+    void startNextScrollRequest() {
+        SearchScrollRequest request = new SearchScrollRequest(mainRequest);
+        request.scrollId(scroll.get()).scroll(firstSearchRequest.scroll());
+        client.searchScroll(request, new ActionListener<SearchResponse>() {
+            @Override
+            public void onResponse(SearchResponse response) {
+                onScrollResponse(response);
+            }
+
+            @Override
+            public void onFailure(Throwable e) {
+                finishHim(e);
+            }
+        });
+    }
+
+    private void recordFailure(Failure failure) {
+        switch (failure.getStatus()) {
+        case CONFLICT:
+            versionConflicts.incrementAndGet();
+            if (mainRequest.abortOnVersionConflict()) {
+                failures.add(failure);
+            }
+            return;
```

Comment:
I usually prefer it in methods that just handle the one thing. But I think I'm going to rewrite this so its more clear. I originally wrote the case statement because I thought there would be more things to handle here. None of them have come up though.